securedrop/source_app/main.py

```diff
@@ -66,53 +71,45 @@ def select_logo() -> werkzeug.Response:
 
     @view.route('/create', methods=['POST'])
     def create() -> werkzeug.Response:
-        if session.get('logged_in', False):
+        if logged_in():
             flash(gettext("You are already logged in. Please verify your codename above as it " +
                           "may differ from the one displayed on the previous page."),
                   'notification')
         else:
             tab_id = request.form['tab_id']
             codename = session['codenames'][tab_id]
-            session['codename'] = codename
-
             del session['codenames']
 
-            filesystem_id = current_app.crypto_util.hash_codename(codename)
             try:
-                source = Source(filesystem_id, current_app.crypto_util.display_id())
-            except ValueError as e:
-                current_app.logger.error(e)
-                flash(
-                    gettext("There was a temporary problem creating your account. "
-                            "Please try again."
-                            ),
-                    'error'
+                current_app.logger.info("Creating new source user...")
+                create_source_user(
+                    db_session=db.session,
+                    source_passphrase=codename,
+                    source_app_crypto_util=current_app.crypto_util,
+                    source_app_storage=current_app.storage,
                 )
-                return redirect(url_for('.index'))
-
-            db.session.add(source)
-            try:
-                db.session.commit()
-            except IntegrityError as e:
-                db.session.rollback()
-                current_app.logger.error(
-                    "Attempt to create a source with duplicate codename: %s" %
-                    (e,))
-
-                # Issue 2386: don't log in on duplicates
-                del session['codename']
+            except (SourcePassphraseCollisionError, SourceDesignationCollisionError) as e:
+                current_app.logger.error("Could not create a source: {}".format(e))
```

Comment:
This consolidates the app's behavior when something goes wrong during account creation.